nep-12.mediawiki

```diff
@@ -0,0 +1,2130 @@
+  switch(type) {
+    case NO_PROVIDER:
+      console.log('No provider available.');
+      break;
+    case CONNECTION_REFUSED:
+      console.log('Connection dApp not connected. Please call the "connect" function.');
+      break;
+    case RPC_ERROR:
+      console.log('There was an error when broadcasting this transaction to the network.');
+      break;
+  }
+});
+</pre>
+
+
+'''Provider Request Handling'''
+
+Upon receiving this request
+* Validate and format the inputs
+** Script hash is valid
+* Broadcast the RPC request
+* Return the RPC response results
+
+Example
+<pre>
+{
+  result: 'hello world'
+}
+</pre>
+
+
+====invokeRead====
+Execute a contract invocation in read-only mode.
+
+
+'''Method Interface'''
+
+<pre>
+function invokeRead({
+  scriptHash: string,
+  operation: string,
+  args: Argument[]
+  network?: string,
+ }): Promise<result: Object>
+</pre>
+
+'''Input arguments'''
+
+<pre>
+interface InvokeReadArgs {
+  scriptHash: string; // script hash of the smart contract to invoke a read on
+  operation: string; // operation on the smart contract to call
+  args: Argument[]; // any input arguments for the operation
+  network?: string - Network to submit this request to. If omitted, will default to network the wallet is currently set to.
+}
+
+interface Argument {
+  type: ArgumentDataType;
+  value: any;
+}
+
+type ArgumentDataType = 'String'|'Boolean'|'Hash160'|'Hash256'|'Integer'|'ByteArray'|'Array'|'Address';
+</pre>
+
+
+'''Success return value'''
+
+result: Object - Response from RPC node
+
+
+'''Error return value'''
+
+<pre>
+interface Error {
+  type: string; // `NO_PROVIDER`|`CONNECTION_REFUSED`|`RPC_ERROR`
+  description: string;
+  data: string;
+}
+</pre>
+
+
+'''Example'''
+
+<pre>
+invokeRead({
+  scriptHash: '505663a29d83663a838eee091249abd167e928f5',
+  operation: 'calculatorAdd',
+  arguments: [
+    {
+      type: 'integer',
+      value: 2
+    },
+    {
+      type: 'integer',
+      value: 10
+    }
+  ]
+})
+.then((result: Object) => {
+  console.log('Read invocation result: ' + JSON.stringigy(result));
+})
+.catch(({type: string, description: string, data: any}) => {
+  switch(type) {
+    case NO_PROVIDER:
+      console.log('No provider available.');
+      break;
+    case CONNECTION_REFUSED:
+      console.log('Connection dApp not connected. Please call the "connect" function.');
+      break;
+   case RPC_ERROR:
+    console.log('There was an error when broadcasting this transaction to the network.');
+    break;
+  }
+});
+</pre>
+
+
+'''Provider Request Handling'''
+
+Upon receiving this request
+* Validate and format the inputs
+** Script hash is valid
+** Format parameters
+*** If `byteArray`, convert to appropriate hexstring based on assumed format
+* Broadcast the RPC request
+* Return the RPC response results
+
+Example
+<pre>
+{
+  script: '8h89fh398f42f.....89hf2894hf9834',
+  state: 'HALT, BREAK',
+  gas_consumed: '0.13',
+  stack: [
+    {
+      type: 'Integer',
+      value: '1337'
```

Comment:
Value is always string, even for integer types? I guess it should respect the type, no?
Integer -> int
Boolean -> bool
ByteArray -> string (hexstring)
etc

Reply:
This was a request from another developer to have all values as type string to maintain consistency in strongly typed languages. I have no preference either way.

Reply:
Good for me, although Array, Struct and Map should exceptions to this,I guess. But as long as this is clearly indicated thats fine (VM stack items will be represented as string, except for these types... example,Integer "123", boolean "true" ??? ) good to clarify this.